external-crates/move/crates/move-vm-runtime/src/tracing2/tracer.rs

```diff
@@ -0,0 +1,1812 @@
+                type_: self.tag.clone(),
+                ref_type: self.layout.1.clone(),
+            }
+        }
+    }
+
+    impl RuntimeLocation {
+        fn into_trace_location(&self) -> Location {
+            match self {
+                RuntimeLocation::Stack(_) => {
+                    panic!("Cannot convert stack location to trace location")
+                }
+                RuntimeLocation::Local(fidx, lidx) => Location::Local(*fidx, *lidx),
+                RuntimeLocation::Indexed(loc, idx) => {
+                    Location::Indexed(Box::new(loc.into_trace_location()), *idx)
+                }
+                RuntimeLocation::Global(id) => Location::Global(*id),
+            }
+        }
+
+        fn into_runtime_location(loc: Location) -> Self {
+            match loc {
+                Location::Local(fidx, lidx) => RuntimeLocation::Local(fidx, lidx),
+                Location::Indexed(loc, idx) => RuntimeLocation::Indexed(
+                    Box::new(RuntimeLocation::into_runtime_location(*loc)),
+                    idx,
+                ),
+                Location::Global(id) => RuntimeLocation::Global(id),
+            }
+        }
+    }
+
+    impl LocalType {
+        fn into_rooted_type(self) -> Option<RootedType> {
+            let ref_type = match self.ref_type {
+                ReferenceType::Value => None,
+                ReferenceType::Empty { .. } => panic!("Empty reference type"),
+                ReferenceType::Filled { ref_type, location } => Some((ref_type, location)),
+            };
+            Some(RootedType {
+                layout: self.layout?,
+                ref_type,
+            })
+        }
+    }
+
+    impl RootedType {
+        fn into_local_type(self) -> LocalType {
+            let ref_type = match self.ref_type {
+                None => ReferenceType::Value,
+                Some((ref_type, location)) => ReferenceType::Filled { ref_type, location },
+            };
+            LocalType {
+                layout: Some(self.layout),
+                ref_type,
+            }
+        }
+    }
+
+    impl<'a> VMTracer<'a> {
+        /// Unwrap or emit an error event to the trace
+        fn unwrap_or_emit_err(&mut self, x: Option<()>) {
+            x.unwrap_or_else(|| {
+                self.trace.effect(EF::ExecutionError(
+                    "!! TRACING ERROR !! Events below this may be incorrect.".to_string(),
+                ));
+            });
+        }
+
+        /// Get the current locals type and reference state(s)
+        fn current_frame_locals(&self) -> Option<&[LocalType]> {
+            Some(
+                self.active_frames
+                    .last_key_value()?
+                    .1
+                    .locals_types
+                    .as_slice(),
+            )
+        }
+
+        /// Return the current frame identifier. This is trace index of the frame and is used to
+        /// identify reference locations rooted higher up the call stack.
```

Comment:
should any of those be a doc comment instead of just a comment?

Reply:
I think doc comments are fine? Can move to normal comments if you want, but figure this is a bit nicer.